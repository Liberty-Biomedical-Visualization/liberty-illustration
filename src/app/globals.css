@tailwind base;
@tailwind components;
@tailwind utilities;

:root {
  --accent-50: 224 247 252;
  --accent-100: 178 224 240;
  --accent-200: 128 200 232;
  --accent-300: 77 176 215;
  --accent-400: 26 153 206;
  --accent-500: 3 152 206;
  --accent-600: 2 128 165;
  --accent-700: 2 109 136;
  --accent-800: 2 90 107;
  --accent-900: 2 61 77;
  --accent-950: 1 44 58;
  --background: 255 255 255;
  --foreground: 23 23 23;
}

body {
  color: var(--foreground);
  background: var(--background);
  font-family: Arial, Helvetica, sans-serif;
}

@layer utilities {
  .text-balance {
    text-wrap: balance;
  }
}
