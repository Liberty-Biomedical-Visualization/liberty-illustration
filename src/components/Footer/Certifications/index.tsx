import resolveClassNames from "@/lib/resolveClassNames";
import AmiMember, { type AmiMemberProps } from "./AmiMember";
import CmiCertification, {
  type CmiCertificationProps,
} from "./CmiCertification";

export default function Certifications(props: Readonly<CertificationsProps>) {
  const { amiLogoData, cmiLogoData } = props;
  const className = resolveClassNames(
    "flex flex-col gap-x-8 gap-y-4 sm:flex-row",
    props.className,
  );

  return (
    <ul className={className}>
      <li>
        <CmiCertification logoData={cmiLogoData} />
      </li>
      <li>
        <AmiMember logoData={amiLogoData} />
      </li>
    </ul>
  );
}

export interface CertificationsProps {
  amiLogoData: AmiMemberProps["logoData"];
  className?: string;
  cmiLogoData: CmiCertificationProps["logoData"];
}
