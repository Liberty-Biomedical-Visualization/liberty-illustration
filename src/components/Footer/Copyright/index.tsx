export default function Copyright(props: Readonly<CopyrightProps>) {
  const { currentYear, holder, initialYear } = props;
  const rangeToCurrentYear = currentYear > initialYear ? `–${currentYear}` : "";
  const content = `© ${initialYear}${rangeToCurrentYear} ${holder}. All rights reserved.`;
  return <small className="text-nowrap">{content}</small>;
}

export interface CopyrightProps {
  readonly currentYear: number;
  readonly holder: string;
  readonly initialYear: number;
}
