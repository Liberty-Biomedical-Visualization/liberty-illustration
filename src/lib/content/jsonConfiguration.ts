/**
 * Expected structure of JSON entries in Contentful.
 */
export const jsonConfiguration = {
  "Home Configuration": {
    carouselImageGalleryId: "string",
    slideDuration: "number",
  },
  "Integration Test JSON": {
    foo: "string",
  },
  "Invalid Integration Test JSON": {
    fizz: "string[]",
    foo: "string",
    zip: "string[]",
  },
  "Medical Illustration Configuration": {
    imageGalleryId: "string",
  },
  "Portfolio Configuration": {
    imageGalleryIds: "string[]",
  },
  "Site Configuration": {
    amiLogoId: "string",
    cmiLogoId: "string",
    copyrightYear: "number",
    navPageOrder: "string[]",
    siteLogoId: "string",
  },
  "Site Metadata": {
    author: "string",
    description: "string",
    faviconId: "string",
    keywords: "string[]",
    locale: "string",
    name: "string",
    url: "string",
  },
} as const;

export type JsonTitle = keyof JsonConfiguration;
type JsonConfiguration = typeof jsonConfiguration;

export type Json<K extends JsonTitle> = StringLiteralPropertiesAsTypes<
  JsonConfiguration[K]
>;

type StringLiteralPropertiesAsTypes<T extends JsonConfiguration[JsonTitle]> = {
  [K in keyof T]: StringLiteralAsType<T[K]>;
};

type StringLiteralAsType<T> = T extends ConfigurableType
  ? ConfigurableTypeMap[T]
  : never;

type ConfigurableType = keyof ConfigurableTypeMap;

type ConfigurableTypeMap = {
  number: number;
  string: string;
  "string[]": string[];
};
