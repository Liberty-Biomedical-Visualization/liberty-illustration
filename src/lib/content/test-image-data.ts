import config from "../config";
import type { ImageData } from ".";

export const completeImage: Readonly<Required<ImageData>> = {
  description: "A complete image used for testing.",
  height: 100,
  src: "https://foo.com/complete.jpg",
  title: "Complete Image",
  width: 100,
};

export const describedImage: Readonly<ImageData & { description: string }> = {
  description: "A described image used for testing.",
  height: 100,
  src: "https://foo.com/described.jpg",
  width: 100,
};

export const imageA: Readonly<Required<ImageData>> = {
  description: "An image identified as A and used for testing.",
  height: 100,
  src: "https://foo.com/a.jpg",
  title: "Image A",
  width: 100,
};

export const imageB: Readonly<Required<ImageData>> = {
  description: "An image identified as B and used for testing.",
  height: 100,
  src: "https://foo.com/b.jpg",
  title: "Image B",
  width: 100,
};

export const imageC: Readonly<Required<ImageData>> = {
  description: "An image identified as C and used for testing.",
  height: 100,
  src: "https://foo.com/c.jpg",
  title: "Image C",
  width: 100,
};

export const integrationTestImage: Readonly<Required<ImageData>> = {
  description: "This is a dummy image used for integration testing.",
  height: 2_480,
  src: `https://${config.CONTENTFUL_IMAGE_CDN_HOSTNAME}/${config.CONTENTFUL_SPACE_ID}/7ci4trPV2FWGxweCTSVqol/b6e816c40b1cf429be48c82a0f73037c/test-image.png`,
  title: "Integration Test Image",
  width: 2_480,
};

export const longDescribedImage: Readonly<ImageData & { description: string }> =
  {
    description:
      "This Is An Image With An Exceedingly Long Description That Will Be Truncated When Converted to a Thumbnail Because It Is Longer Than the Maximum Character Count",
    height: 100,
    src: "https://foo.com/long-titled.jpg",
    width: 100,
  };

export const longTitledImage: Readonly<ImageData & { title: string }> = {
  height: 100,
  src: "https://foo.com/long-titled.jpg",
  title:
    "An Image With An Exceedingly Long Title That Will Be Truncated When Converted to a Thumbnail Because It Is Longer Than the Maximum Character Count",
  width: 100,
};

export const minimalImage: Readonly<ImageData> = {
  height: 100,
  src: "https://foo.com/minimal.jpg",
  width: 100,
};

export const nondescriptImage: Readonly<ImageData & { title: string }> = {
  height: 100,
  src: "https://foo.com/nondescript.jpg",
  title: "Nondescript Image",
  width: 100,
};

export const titledImage: Readonly<ImageData & { title: string }> = {
  height: 100,
  src: "https://foo.com/titled.jpg",
  title: "Titled Image",
  width: 100,
};

export const untitledImage: Readonly<ImageData & { description: string }> = {
  description: "An untitled image used for testing.",
  height: 100,
  src: "https://foo.com/untitled.jpg",
  width: 100,
};
