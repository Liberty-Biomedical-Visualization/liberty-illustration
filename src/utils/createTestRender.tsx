import { render } from "@testing-library/react";
import type { JSX, JSXElementConstructor } from "react";

/**
 * Returns a function that allows rendering Component with partial props while
 * falling back to defaultProps for testing purposes.
 */
export default function createTestRender<C extends JSXElementConstructor<P>, P>(
  Component: JSXElementConstructor<P> & C,
  defaultProps: JSX.LibraryManagedAttributes<C, P>,
) {
  return function (props: Partial<P> = {}) {
    const resolvedProps: JSX.IntrinsicAttributes &
      JSX.LibraryManagedAttributes<C, P> = { ...defaultProps, ...props };
    return render(<Component {...resolvedProps} />);
  };
}
