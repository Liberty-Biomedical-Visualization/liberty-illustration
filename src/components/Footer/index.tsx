import resolveClassNames from "@/lib/resolveClassNames";

import Certifications, { type CertificationsProps } from "./Certifications";
import Copyright, { type CopyrightProps } from "./Copyright";

export default function Footer(props: Readonly<FooterProps>) {
  const {
    amiLogoData,
    cmiLogoData,
    copyrightHolder,
    copyrightYear,
    currentYear,
  } = props;

  const className = resolveClassNames(
    "flex flex-col gap-4 items-center",
    props.className,
  );

  return (
    <footer className={className}>
      <Certifications amiLogoData={amiLogoData} cmiLogoData={cmiLogoData} />
      <Copyright
        currentYear={currentYear}
        holder={copyrightHolder}
        initialYear={copyrightYear}
      />
    </footer>
  );
}

export interface FooterProps {
  amiLogoData: CertificationsProps["amiLogoData"];
  className?: string;
  cmiLogoData: CertificationsProps["cmiLogoData"];
  copyrightHolder: CopyrightProps["holder"];
  copyrightYear: CopyrightProps["initialYear"];
  currentYear: CopyrightProps["currentYear"];
}
