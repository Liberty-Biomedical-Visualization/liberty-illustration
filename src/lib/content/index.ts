import * as contentful from "contentful";

import packageData from "../../../package.json";
import config from "../config";
import isDefined from "../isDefined";
import { Json, JsonTitle } from "./jsonConfiguration";
import { validateJson } from "./validateJson";

/**
 * Gets `AssetData` by ID.
 *
 * Throws an `Error` if the asset is not found.
 */
export async function getAssetData(id: string) {
  const asset = await client.getAsset(id);
  return mapContentfulAssetToAssetData(asset);
}

export interface AssetData {
  description?: string;
  title?: string;
  url: string;
}

/**
 * Gets `ImageData` by ID.
 *
 * Throws an `Error` if the image is not found, or the asset is not an image.
 */
export async function getImageData(id: string) {
  const asset = await client.getAsset(id);
  return mapContentfulAssetToImageData(asset);
}

export interface ImageData {
  description?: string;
  height: number;
  src: string;
  title?: string;
  width: number;
}

/**
 * Gets an `ImageGallery` by ID.
 *
 * Throws an `Error` if the gallery is not found.
 */
export async function getImageGallery(id: string) {
  const entry = await client.getEntry<ImageGallerySkeleton>(id);
  return mapEntryToImageGallery(entry);
}

export interface ImageGallery {
  images: ImageData[];
  title: string;
}

/**
 * Gets a `JsonReference` by title.
 *
 * Throws an error if the JSON is not found or not properly structured.
 */
export async function getJsonByTitle<K extends JsonTitle>(title: K) {
  const jsons = await client.getEntries<JsonSkeleton<Json<K>>>({
    content_type: "json",
    "fields.title": title,
  });
  const json = jsons.items[0];

  if (json === undefined) {
    throw new Error(`JSON with title "${title}" not found`);
  }

  validateJson(json.fields.value, title);
  return json.fields.value;
}

/**
 * Gets a `VideoGallery` by ID.
 *
 * Throws an `Error` if the gallery is not found.
 */
export async function getVideoGallery(id: string) {
  const entry = await client.getEntry<VideoGallerySkeleton>(id);
  return mapEntryToVideoGallery(entry);
}

export interface VideoGallery {
  title: string;
  videos: VideoData[];
}

export interface VideoData {
  title: string;
  url: string;
}

function mapContentfulAssetToAssetData(asset: ContentfulAsset) {
  const { description, file, title } = asset.fields;

  if (file === undefined) {
    throw new Error("file is undefined");
  }

  const assetData: AssetData = { url: "https:" + file.url };

  if (description) {
    assetData.description = description;
  }

  if (title) {
    assetData.title = title;
  }

  return assetData;
}

function mapContentfulAssetToImageData(asset: ContentfulAsset) {
  const { file } = asset.fields;

  if (file === undefined) {
    throw new Error("file is undefined");
  }

  const { image } = file.details;

  if (image === undefined) {
    throw new Error("image is undefined");
  }

  const { height, width } = image;
  const { description, title, url } = mapContentfulAssetToAssetData(asset);
  const imageData: ImageData = { height, src: url, width };

  if (description) {
    imageData.description = description;
  }

  if (title) {
    imageData.title = title;
  }

  return imageData;
}

type ContentfulAsset = contentful.Asset<"WITHOUT_UNRESOLVABLE_LINKS", string>;

function mapEntryToImageGallery(entry: ImageGalleryEntry) {
  const { images, title } = entry.fields;
  const mappedImages = images
    .filter(isDefined)
    .map(mapContentfulAssetToImageData);
  const gallery: ImageGallery = { images: mappedImages, title };
  return gallery;
}

type ImageGalleryEntry = contentful.Entry<
  ImageGallerySkeleton,
  "WITHOUT_UNRESOLVABLE_LINKS",
  string
>;

interface ImageGallerySkeleton {
  contentTypeId: "imageGallery";
  fields: {
    images: contentful.EntryFieldTypes.Array<contentful.EntryFieldTypes.AssetLink>;
    title: contentful.EntryFieldTypes.Text;
  };
}

function mapEntryToVideoGallery(entry: VideoGalleryEntry) {
  const { title, videos } = entry.fields;
  const mappedVideos = videos.filter(isDefined).map((video) => video.fields);
  const gallery: VideoGallery = { title, videos: mappedVideos };
  return gallery;
}

type VideoGalleryEntry = contentful.Entry<
  VideoGallerySkeleton,
  "WITHOUT_UNRESOLVABLE_LINKS",
  string
>;

interface VideoGallerySkeleton {
  contentTypeId: "imageGallery";
  fields: {
    title: contentful.EntryFieldTypes.Text;
    videos: contentful.EntryFieldTypes.Array<
      contentful.EntryFieldTypes.EntryLink<VideoLinkSkeleton>
    >;
  };
}

interface VideoLinkSkeleton {
  contentTypeId: "videoLink";
  fields: {
    title: contentful.EntryFieldTypes.Text;
    url: contentful.EntryFieldTypes.Text;
  };
}

interface JsonSkeleton<T extends JsonReference> {
  contentTypeId: "json";
  fields: {
    value: contentful.EntryFieldTypes.Object<T>;
    title: contentful.EntryFieldTypes.Text;
  };
}

type JsonReference = JsonArray | JsonObject;
type JsonArray = JsonValue[];
type JsonObject = { [key: string]: JsonValue };
type JsonValue = JsonArray | JsonObject | boolean | null | number | string;

const { dependencies, name, version } = packageData;

const client = contentful.createClient({
  accessToken: config.CONTENTFUL_ACCESS_TOKEN,
  application: `${name}/${version}`,
  environment: config.CONTENTFUL_ENVIRONMENT,
  integration: `next/${dependencies.next}`,
  space: config.CONTENTFUL_SPACE_ID,
}).withoutUnresolvableLinks;
