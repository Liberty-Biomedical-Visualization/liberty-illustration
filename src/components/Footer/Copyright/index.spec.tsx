import { describe, expect, it } from "@jest/globals";
import { screen } from "@testing-library/react";

import createTestRender from "@/lib/createTestRender";

import Copyright from ".";

describe("Copyright", () => {
  it("should display initialYear when currentYear is equal to initialYear", () => {
    const holder = "John Doe";
    const year = 2024;
    renderCopyright({ currentYear: year, holder, initialYear: year });

    const expectedText = `© ${year} ${holder}. All rights reserved.`;
    const copyright = screen.getByText(expectedText);
    expect(copyright).toBeVisible();
  });

  it("should display a range of dates when currentYear is greater than initialYear", () => {
    const holder = "Jane Doe";
    const initialYear = 2023;
    const currentYear = initialYear + 1;
    renderCopyright({ currentYear, holder, initialYear });

    const expectedText = `© ${initialYear}–${currentYear} ${holder}. All rights reserved.`;
    const copyright = screen.getByText(expectedText);
    expect(copyright).toBeVisible();
  });

  it("should display initialYear when currentYear is less than initialYear", () => {
    const holder = "Jean Doe";
    const initialYear = 2024;
    const currentYear = initialYear - 1;
    renderCopyright({ currentYear, holder, initialYear });

    const expectedText = `© ${initialYear} ${holder}. All rights reserved.`;
    const copyright = screen.getByText(expectedText);
    expect(copyright).toBeVisible();
  });

  it("should match the snapshot", () => {
    const result = renderCopyright();
    expect(result.container).toMatchSnapshot();
  });
});

const renderCopyright = createTestRender(Copyright, {
  currentYear: 2024,
  holder: "John Doe",
  initialYear: 2023,
});
