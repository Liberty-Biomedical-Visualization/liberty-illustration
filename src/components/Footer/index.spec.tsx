import { describe, expect, it } from "@jest/globals";
import { screen } from "@testing-library/dom";

import { completeImage, minimalImage } from "@/lib/content/test-image-data";
import createTestRender from "@/lib/createTestRender";

import Footer from ".";

describe("Footer", () => {
  it("should display the site content info", () => {
    renderFooter();
    const result = screen.getByRole("contentinfo");
    expect(result).toBeVisible();
  });

  it("should display the AMI logo", () => {
    renderFooter({ amiLogoData: completeImage });
    const result = screen.getByRole("img", { name: completeImage.description });
    expect(result).toBeVisible();
  });

  it("should display the CMI logo", () => {
    renderFooter({ cmiLogoData: completeImage });
    const result = screen.getByRole("img", { name: completeImage.description });
    expect(result).toBeVisible();
  });

  it("should display the copyright holder", () => {
    const copyrightHolder = "Foo Bar";
    renderFooter({ copyrightHolder });
    const result = screen.getByText(copyrightHolder, { exact: false });
    expect(result).toBeVisible();
  });

  it("should display the copyright year", () => {
    const copyrightYear = 1999;
    renderFooter({ copyrightYear });
    const result = screen.getByText(copyrightYear, { exact: false });
    expect(result).toBeVisible();
  });

  it("should display the current year", () => {
    const currentYear = 2077;
    renderFooter({ currentYear });
    const result = screen.getByText(currentYear, { exact: false });
    expect(result).toBeVisible();
  });

  it("should pass className to the outermost element", () => {
    const className = "foo";
    const result = renderFooter({ className });
    expect(result.container.firstChild).toHaveClass(className);
  });

  it("should match the snapshot", () => {
    const result = renderFooter();
    expect(result.container).toMatchSnapshot();
  });
});

const renderFooter = createTestRender(Footer, {
  amiLogoData: minimalImage,
  cmiLogoData: minimalImage,
  copyrightHolder: "Abigail Richbourg Liberty",
  copyrightYear: 2014,
  currentYear: 2024,
});
