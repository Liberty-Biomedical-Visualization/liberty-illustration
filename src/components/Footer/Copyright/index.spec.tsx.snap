// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Copyright should match the snapshot 1`] = `
<div>
  <small
    class="text-nowrap"
  >
    © 2023–2024 John Doe. All rights reserved.
  </small>
</div>
`;
